source/slang/slang-ir-sccp.cpp

```diff
@@ -288,10 +793,116 @@ struct SCCPContext
         // `None` inputs as producing `Any` to make sure we don't
         // optimize the code based on non-obvious assumptions.
         //
-        // For now we aren't implementing *any* folding logic here,
-        // for simplicity. This is the right place to add folding
-        // optimizations if/when we need them.
-        //
+        // For now we implement only basic folding operations for
+        // scalar values.
+        if (!as<IRBasicType>(inst->getDataType()))
+            return LatticeVal::getAny();
+
+        switch (inst->getOp())
+        {
+        case kIROp_Construct:
+            return evalConstruct(inst->getFullType(), getLatticeVal(inst->getOperand(0)));
+        case kIROp_Add:
+            return evalAdd(
+                inst->getFullType(),
```

Comment:
Note: using `getFullType` here might lead to issues, since the distinction between `getType()` and `getFullType()` is that `getFullType()` can return a rate-qualified type. We use rate-qualified types *very* rarely in the IR right now - basically just as our way to encode `groupshared` variables (`@GroupShared`) and `constexpr` (`@ConstExpr`).

I suspect that in practice we actually just want `getType()` for all of these, but I can't be entirely certain; if we ended up using rates for more stuff down the line we might decide that the folding logic here needs to take them into account.